File: node/src/crypto/nonce.ts
import { Bytes, Uint64 } from './types';

export class Nonce {
    private n: Uint64
    private readonly bytes: Bytes
    private readonly view: DataView

    constructor (size: number) {
        this.n = 0n
        this.bytes = new Uint8Array(size);
        this.view = new DataView(this.bytes.buffer, this.bytes.byteOffset, this.bytes.byteLength);
        this.view.setBigUint64(size - 8, this.n, true);
    }

    increment (): void {
        this.n++;
        // Even though we're treating the nonce as 8 bytes, RFC7539 specifies 12 bytes for a nonce.
        this.view.setBigUint64(this.bytes.length - 8, this.n, true);
    }

    getBytes (): Bytes {
        return this.bytes;
    }

    getUint64 (): Uint64 {
        return this.n;
    }
}

File: go/sessionstate/nonce.go
package sessionstate

import "encoding/binary"

type Nonce struct {
	n     uint64
	bytes []byte
}

// NewNonce creates a new Nonce with the specified size
func NewNonce(size int) *Nonce {
	n := &Nonce{
		n:     0,
		bytes: make([]byte, size),
	}
	// Even though we're treating the nonce as 8 bytes, RFC7539 specifies 12 bytes for a nonce.
	binary.LittleEndian.PutUint64(n.bytes[size-8:], n.n)
	return n
}

func (n *Nonce) Increment() {
	n.n++
	binary.LittleEndian.PutUint64(n.bytes[len(n.bytes)-8:], n.n)
}

// GetBytes returns the byte representation of the nonce
func (n *Nonce) GetBytes() []byte {
	return n.bytes
}

// GetUint64 returns the counter value as uint64
func (n *Nonce) GetUint64() uint64 {
	return n.n
}
